spark/ingestion/src/test/scala/feast/ingestion/BatchPipelineIT.scala

```diff
@@ -0,0 +1,306 @@
+        )
+      )
+    })
+  }
+
+  "Old rows in ingestion" should "not overwrite more recent rows from storage" in new Scope {
+    val entities = (0 to 10000).map(_.toString)
+
+    val genLatest = rowGenerator(
+      DateTime.parse("2020-08-15"),
+      DateTime.parse("2020-09-01"),
+      Some(Gen.oneOf(entities))
+    )
+    val latest = generateDistinctRows(genLatest, 10000)
+
+    val tempPath1 = storeAsParquet(latest)
+    val config1   = config.copy(source = FileSource(tempPath1, Map.empty, "eventTimestamp"))
+
+    BatchPipeline.createPipeline(sparkSession, config1)
+
+    val genOld = rowGenerator(
+      DateTime.parse("2020-08-01"),
+      DateTime.parse("2020-08-14"),
+      Some(Gen.oneOf(entities))
+    )
+    val old = generateDistinctRows(genOld, 10000)
+
+    val tempPath2 = storeAsParquet(old)
+    val config2   = config.copy(source = FileSource(tempPath2, Map.empty, "eventTimestamp"))
+
+    BatchPipeline.createPipeline(sparkSession, config2)
+
+    val featureKeyEncoder: String => String = encodeFeatureKey(config.featureTable)
+
+    latest.foreach(r => {
+      val storedValues = jedis.hgetAll(encodeEntityKey(r, config.featureTable)).asScala.toMap
+      storedValues should beStoredRow(
+        Map(
+          featureKeyEncoder("feature1") -> r.feature1,
+          featureKeyEncoder("feature2") -> r.feature2,
+          "_ts:test-fs"                 -> r.eventTimestamp
+        )
+      )
+    })
+  }
+
+  "Invalid rows" should "not be ingested and stored to deadletter instead" in new Scope {
+    val gen  = rowGenerator(DateTime.parse("2020-08-01"), DateTime.parse("2020-09-01"))
+    val rows = generateDistinctRows(gen, 100)
+
+    val rowsWithNullEntity = rows.map(_.copy(customer = null))
+
+    val tempPath = storeAsParquet(rowsWithNullEntity)
+    val deadletterConfig = config.copy(
+      source = FileSource(tempPath, Map.empty, "eventTimestamp"),
+      deadLetterPath = Some(generateTempPath("deadletters"))
+    )
+
+    BatchPipeline.createPipeline(sparkSession, deadletterConfig)
+
+    jedis.keys("*").toArray should be(empty)
+
+    sparkSession.read
+      .parquet(deadletterConfig.deadLetterPath.get)
+      .count() should be(rows.length)
+  }
+
+  "Columns from source" should "be mapped according to configuration" in new Scope {
+    val gen  = rowGenerator(DateTime.parse("2020-08-01"), DateTime.parse("2020-09-01"))
+    val rows = generateDistinctRows(gen, 100)
+
+    val tempPath = storeAsParquet(rows)
+
+    val configWithMapping = config.copy(
+      featureTable = config.featureTable.copy(
+        entities = Seq(Field("entity", ValueType.Enum.STRING)),
+        features = Seq(
+          Field("new_feature1", ValueType.Enum.INT32),
+          Field("new_feature2", ValueType.Enum.FLOAT)
+        )
+      ),
+      source = FileSource(
+        tempPath,
+        Map(
+          "entity"       -> "customer",
+          "new_feature1" -> "feature1",
+          "new_feature2" -> "feature2"
```

Comment:
Is it possible to provide a partial list here? Lets say one feature mapping is already 1:1 and the other one requires mapping.

Reply:
Changed test